packages/react-router/modules/Router.js

```diff
@@ -20,6 +20,26 @@ class Router extends React.Component {
     router: PropTypes.object.isRequired
   };
 
+  constructor(props) {
+    super(props);
+
+    const { children, history } = props;
+
+    invariant(
+      children == null || React.Children.count(children) === 1,
+      "A <Router> may have only one child element"
+    );
+
+    // Do this here so we can setState when a <Redirect> changes the
+    // location in componentDidMount. This happens e.g. when doing
+    // server rendering using a <StaticRouter>.
+    this.unlisten = history.listen(() => {
```

Comment:
The listener should be done in a componentDidMount. This will fire during server rendering and even if the component never gets mounted.

Reply:
The challenge with having it in cDM (which is where I had it originally) is that -- from what I found and the tests that failed -- React doesn't guarantee the order of whether a parent or child's cDM is fired first. Which means that the `Redirect` component doesn't work, since the logic for redirection happens in cDM as well.

Maybe I could look at putting it in both places and checking if it's in a static router or something? Not sure. Any ideas?

Reply:
The React blog has some info on this: https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html#adding-event-listeners-or-subscriptions

Reply:
@timdorr from what I'm reading, they recommend using cDM. Yet that doesn't solve the issue I mentioned above.

I pushed up an alternative solution, what do you think of it?